src/AspNet.Security.OAuth.Introspection/Events/ValidateAudienceContext.cs

```diff
@@ -0,0 +1,45 @@
+﻿using Microsoft.AspNetCore.Http;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace AspNet.Security.OAuth.Introspection {
+    public class ValidateAudienceContext : BaseIntrospectionContext {
+        public ValidateAudienceContext(
+            HttpContext context, 
+            OAuthIntrospectionOptions options,
+            JToken payload) 
+            : base(context, options) {
+            Payload = payload;
+            var audiences = payload[OAuthIntrospectionConstants.Claims.Audience];
+            if(audiences != null) {
+                if(audiences.Type == JTokenType.String) {
+                    Audiences = new[] { payload.Value<string>(OAuthIntrospectionConstants.Claims.Audience) };
+                }
+                else if (audiences.Type == JTokenType.Array) {
+                    Audiences = payload.Value<JArray>(OAuthIntrospectionConstants.Claims.Audience)
+                                           .Select(audience => audience.Value<string>());
+                }
+            }
+            Validate();
+        }
+
+        public IEnumerable<string> Audiences { get; set; }
+
+        public JToken Payload { get; }
+
+        public bool IsValid { get; set; }
+
+        public bool Validate() {
+            return IsValid = Validate(Audiences);
+        }
+
+        public bool Validate(IEnumerable<string> audiences)
+        {
+            IsValid = (audiences == null && Options.Audiences.Count > 0) || Options.Audiences.Count == 0 ||
```

Comment:
Well, in this case, I see three potential collections of audiences - ones configured on Options, ones in the `AuthenticationTicket`, and ones that a dev overrides with. I thought it would be convenient to have the context check whichever collection you want against that configured in Options.

Related: I almost separated out the default logic into the context itself for separation of concerns, but decided against that because I'm not sure what the impact of that would be. (Though making the middleware's default behavior actually in the contexts mainly might make the handler much easier to work with... Not sure about this, but it's something to think about.)